syncapi/storage/postgres/relations_table.go

```diff
@@ -0,0 +1,199 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package postgres
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/matrix-org/dendrite/internal"
+	"github.com/matrix-org/dendrite/internal/sqlutil"
+	"github.com/matrix-org/dendrite/syncapi/storage/tables"
+	"github.com/matrix-org/dendrite/syncapi/types"
+)
+
+const relationsSchema = `
+CREATE SEQUENCE IF NOT EXISTS syncapi_relation_id;
+
+CREATE TABLE IF NOT EXISTS syncapi_relations (
+	id BIGINT PRIMARY KEY DEFAULT nextval('syncapi_relation_id'),
+	room_id TEXT NOT NULL,
+	event_id TEXT NOT NULL,
+	child_event_id TEXT NOT NULL,
+	rel_type TEXT NOT NULL,
+	CONSTRAINT syncapi_relations_unique UNIQUE (room_id, event_id, child_event_id, rel_type)
+);
+`
+
+const insertRelationSQL = "" +
+	"INSERT INTO syncapi_relations (" +
+	"  room_id, event_id, child_event_id, rel_type" +
+	") VALUES ($1, $2, $3, $4) " +
+	" ON CONFLICT ON CONSTRAINT syncapi_relations_unique DO UPDATE SET event_id=EXCLUDED.event_id" +
+	" RETURNING id"
+
+const deleteRelationSQL = "" +
+	"DELETE FROM syncapi_relations WHERE room_id = $1 AND child_event_id = $2"
+
+const selectRelationsInRangeAscSQL = "" +
+	"SELECT id, room_id, child_event_id, rel_type FROM syncapi_relations" +
+	" WHERE room_id = $1 AND event_id = $2 AND id > $3 AND id <= $4" +
+	" ORDER BY id ASC LIMIT $5"
+
+const selectRelationsInRangeDescSQL = "" +
+	"SELECT id, room_id, child_event_id, rel_type FROM syncapi_relations" +
+	" WHERE room_id = $1 AND event_id = $2 AND id >= $3 AND id < $4" +
+	" ORDER BY id DESC LIMIT $5"
+
+const selectRelationsByTypeInRangeAscSQL = "" +
+	"SELECT id, room_id, child_event_id, rel_type FROM syncapi_relations" +
+	" WHERE room_id = $1 AND event_id = $2 AND rel_type = $3 AND id > $4 AND id <= $5" +
+	" ORDER BY id ASC LIMIT $6"
+
+const selectRelationsByTypeInRangeDescSQL = "" +
+	"SELECT id, room_id, child_event_id, rel_type FROM syncapi_relations" +
+	" WHERE room_id = $1 AND event_id = $2 AND rel_type = $3 AND id >= $4 AND id < $5" +
+	" ORDER BY id DESC LIMIT $6"
+
+const selectMaxRelationIDSQL = "" +
+	"SELECT MAX(id) FROM syncapi_relations"
+
+type relationsStatements struct {
+	insertRelationStmt                   *sql.Stmt
+	selectRelationsInRangeAscStmt        *sql.Stmt
+	selectRelationsInRangeDescStmt       *sql.Stmt
+	selectRelationsByTypeInRangeAscStmt  *sql.Stmt
+	selectRelationsByTypeInRangeDescStmt *sql.Stmt
+	deleteRelationStmt                   *sql.Stmt
+	selectMaxRelationIDStmt              *sql.Stmt
+}
+
+func NewPostgresRelationsTable(db *sql.DB) (tables.Relations, error) {
+	s := &relationsStatements{}
+	_, err := db.Exec(relationsSchema)
+	if err != nil {
+		return nil, err
+	}
+	if s.insertRelationStmt, err = db.Prepare(insertRelationSQL); err != nil {
```

Comment:
Probably best to use `sqlutil.StatementList{}` to prepare the statements, as we do in other places already.